### ISSUE TITLE
Contrast limits slider doesn't update correctly after reset

### ISSUE THREAD

**Issue by jnahlers** [napari/napari#6522]
### 🐛 Bug Report

I am not sure what the root cause is, but I am experiencing that float data with a large range of contrast can cause problems with the contrast slider pop-up. When the range in a single slice is extremely small (e.g. 1E-25 to 4E-25), but the range in the whole array is large (e.g. 1E-25 to 1E25), I constantly need to reset the contrast slider when scrolling through slices, in order to be able to adjust the contrast limits in any one slice. I therefore often open the contrast limits slider pop-up, and hit the 'reset' button. However, after hitting the reset button the slider in the pop-up freezes up and doesn't correctly show the limits in its labels.

### 💡 Steps to Reproduce

```
# imports
import numpy as np
import napari

# create image array
arr = np.arange(1,9,1,dtype=np.float32).reshape((2,2,2))
arr[0,:,:] = arr[0,:,:]*1E-20
arr[1,:,:] = arr[1,:,:]*1E+20

# open the image
viewer, image_layer = napari.imshow(arr)
```

1. Click on 'continuous' auto-contrast.
2. Right-click the contrast limits slider to bring up the contrast limits pop-up.
3. Click the 'reset' button.
4. After clicking the reset button, the slider labels all just show '0', and it is not possible to drag the slider handles.
5. Click in the main window to close the contrast slider pop-up.
6. Again, right-click the contrast limits slider to bring up the contrast limits pop-up.
7. Now the labels show the 'correct' values (0.000000000000... which gets cut off), and it is possible to move the slider handles.
8. Close the pop-up, switch to the second index (1,:,:), and switch back to the first. This 're-sets' the problem. Go back to step 2 and repeat. 

### 💡 Expected Behavior

After hitting the 'reset' button, the slider in the pop-up should update correctly, show the new limits in the labels, and be draggable. 

### 🌎 Environment

napari: 0.5.0a2.dev486+g4d60a7ce
Platform: Windows-10-10.0.19045-SP0
Python: 3.9.18 | packaged by conda-forge | (main, Aug 30 2023, 03:40:31) [MSC v.1929 64 bit (AMD64)]
Qt: 5.15.2
PyQt5: 5.15.10
NumPy: 1.26.2
SciPy: 1.11.4
Dask: 2023.12.0
VisPy: 0.14.1
magicgui: 0.8.0
superqt: 0.6.1
in-n-out: 0.1.9
app-model: 0.2.2
npe2: 0.7.3

OpenGL:
  - GL version:  4.6.0 - Build 27.20.100.9565
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen 1: resolution 1920x1200, scale 1.0

Settings path:
  - C:\Users\...\AppData\Local\napari\napari-env_4624f08893c59545bcb98518a8c42a23d61d4ece\settings.yaml

### 💡 Additional Context

Somewhat unrelated, but I would also suggest that it might be an idea to have the reset button for the contrast limits slider somewhere in the main GUI, as it's a bit of a pain to have to constantly open the pop-up to reset when each slice has wildly different contrast limits.         

Finally, we should consider using scientific notation in the contrast limits slider labels when the numbers get very small or large, as its not very helpful to see a cut-off 0.000000000000...

**Comment by Czaki** [comment]
Does setting the auto contrast to a continuous mode does not solve your original problem of reset clicking? 

![obraz](https://github.com/napari/napari/assets/3826210/5ab5c3c8-16fa-4e36-a104-f1c3f27547e7)

**Comment by jnahlers** [comment]
@Czaki No, that adjusts the contrast limits, but (apparently) not the contrast limits slider max/min when the new contrast limits are within the old max/min. But if the max/min are something like 1E-25 and 1E25, and the new dynamic range is something like 1E-25 to 2E-25, you need to reset the slider to be able to adjust within that tight dynamic range. Hopefully the reproduction code clarifies what I mean?

**Comment by Czaki** [comment]
Please describe use case where you need to do such contrast manual alignment. We need to understand it. Adding any new button to the GUI is a really hard decision. The space is limited.

**Comment by jnahlers** [comment]
Sure! The situation would be any where different slices of an Nd array have very different dynamic ranges, and you want to be able to reasonably quickly both scroll through the slices and adjust the contrast limits in any one slice. The idea of adding the reset button (currently found in the contrast limit slider pop-up) to the main GUI is fairly minor QoL matter - a keyboard shortcut to reset the contrast limits slider max/min to the dynamic range of the current slice would work just as well.

All this is not, however, the main point of this bug report. The bug is simply that the contrast limits slider (in the pop-up) does not update correctly when the reset button is clicked.

**Comment by psobolewskiPhD** [comment]
> No, that adjusts the contrast limits, but (apparently) not the contrast limits slider max/min when the new contrast limits are within the old max/min.

Ah, OK i get what triggered the issue then. Continuous was working correctly. Maybe it *should* also reset the limits of the slider for some dtypes?

**Comment by Czaki** [comment]
> Ah, OK i get what triggered the issue then. Continuous was working correctly. Maybe it _should_ also reset the limits of the slider for some dtypes?

I rather think that we should switch (or allow to switch) to log sliders.

**Comment by Czaki** [comment]
@jnahlers I have created bug fix PR #6523, and I have raised your suggestion about scientific notation in upstream project https://github.com/pyapp-kit/superqt/issues/224. 

Also, allowing for log scale sliders will require changes in superqt.

**Comment by jnahlers** [comment]
@Czaki Thank you! 

@psobolewskiPhD I have thought a bit more about this. The question essentially is, how should the contrast limits slider actually behave when in continuous auto-contrast mode? Currently, when the new data range is within the min/max range of the contrast limits slider, the min/max doesn't change, leading to the problem that a very tight data range makes the slider unusable (until the min/max is reset by the button in the pop-up).

If the min/max of the slider was continuously changed to the data range while continuous auto-contrast is on, the sliders would be at the ends of the bar all the time, ready to be used.  

A possible downside of this is that contrast limits slider would no longer function as a 'pseudo-histogram' when continuous auto-contrast is on. Some users with data that's always within roughly the same range may find it useful that while scrolling through their data the limits sliders move around, giving them a rough idea of the data range within a slice. Constantly updating the min/max of the slider would get rid of this functionality. 

What exactly does it mean for the user to change the contrast limits while in continuous auto-contrast mode? Currently, it is possible to change the contrast limits while in continuous auto-contrast, but as soon as the data changes the contrast limits that one has selected are lost, and even when turning continuous auto-contrast off after manually adjusting the limits (and I think this is a bug) the contrast limits change to the data range. Perhaps the answer is simply to disable interaction with the contrast limits slider (and pop-up) while continuous auto-contrast is on. Then, anytime the user wishes to adjust the contrast manually, they turn continuous auto-contrast off and the sliders are ready to go.
